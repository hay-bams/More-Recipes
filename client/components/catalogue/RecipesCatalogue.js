import React from 'react';
import { Link } from 'react-router-dom';

/**
 * @class RecipeCatalogue
 */
class RecipeCatalogue extends React.Component {
  render() {
    return (
      <div className = "container">
      <div className = "row">
        <div className = "col-sm">
          <h2 className = "text-center catalogue">Recipes Catalogue</h2>
        </div>
      </div>

      <div className = "row">
        <div className = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div className="card recipe-card" style={{border: 'none'}}>
            <img className="card-img-top img-fluid" src="images/recipe5.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div className="card-body">
              <h4 className="card-title">Recipe name</h4>
              <p className="card-text">
                <div className = "row">
                  <a href="#" className = "text-success"><i className="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" className = "text-info"><i className="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" className = "text-danger"><i className="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe2.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe3.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe4.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

        <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe5.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe6.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe5.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe2.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

        <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe3.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe4.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe5.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>

         <div class = "col-12 col-sm-6 col-md-4 col-lg-3 recipes">
          <div class="card recipe-card" style={{border: 'none'}}>
            <img class="card-img-top img-fluid" src="images/recipe6.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div class="card-body">
              <h4 class="card-title">Recipe name</h4>
              <p class="card-text">
                <div class = "row">
                  <a href="#" class = "text-success"><i class="fa fa-thumbs-up col-4" aria-hidden="true"></i>10</a>
                  <a href="#" class = "text-info"><i class="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" class = "text-danger"><i class="fa fa-thumbs-down col-4" aria-hidden="true"></i>3</a>
                </div>
              </p>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>
      </div>
    </div>
    )
  }
}

export default RecipeCatalogue;
