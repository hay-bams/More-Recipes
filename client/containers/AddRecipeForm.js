import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'; 
import { Link } from 'react-router-dom';
import { addRecipe } from '../actions/actions.js';
import { clearError } from '../actions/actions.js';

/**
 * @class AddRecipeForm
 */
class AddRecipeForm extends React.Component {
  constructor(props) {
    super(props);
    this.addRecipe =  this.addRecipe.bind(this);
    this.recipeNum = this.props.recipes.length;
  }

  componentWillMount() {
     this.props.clearError()
  }

  addRecipe(event) {
    event.preventDefault();
    const recipe = {
      title: this.title.value,
      ingredients: this.ingredients.value,
      instructions: this.instructions.value,
      image: 'imageURl'
    }
    this.props.addRecipe(recipe);
  } 

  render() {
    return (
      <div className = "main-userboard-body">
        <div className = "container-fluid">
          <ol className="breadcrumb">
            <li className="breadcrumb-item">
              <a href="#">Dashboard</a>
            </li>
            <li className="breadcrumb-item active">Add Recipe</li>
          </ol>
        </div>
  
        <div className = "container">
          <div className = "row">
            <div className = "col-sm-12 col-md-8 col-lg-6">
           {this.recipeNum = this.props.recipes.length - this.recipeNum === 1 ? 
            this.props.clearError() && <span></span> : ""} 
              <span>{this.props.message.errors ? this.props.message.errors.response.data.message: ""}</span>
                <form onSubmit = {this.addRecipe}>
                <div className = "form-group">
                  <label htmlFor = "food">Recipe Name</label>
                  <input ref = {(input) => this.title = input} type="text" name="title" 
                  className = "form-control" placeholder="Enter Recipe name" />
                </div>
  
                <div className = "form-group">
                  <input type="file" />
                </div>
  
                  <div className = "form-group">
                  <label htmlFor = "ingredient">Ingredients</label>
                  <input  ref = {(input) => this.ingredients = input}ype="text" name="ingredient" 
                  className = "form-control" placeholder="Enter Ingredients" />
                </div>
  
                <div className = "form-group">
                  <label htmlFor = "instruction">Instructions</label>
                  <textarea ref = {(input) => this.instructions = input} className = "form-control" 
                  placeholder="Enter Instructions" name = "instruction"></textarea>
                </div>
  
                <div className = "form-group">
                  <input type="submit" className = "btn btn-success" value = "Add Recipe" />
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
  
    )
  }
}

const mapStateToProps = state => {
  return {
    recipes: state.recipes,
    message: state.errors
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({addRecipe, clearError}, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(AddRecipeForm);