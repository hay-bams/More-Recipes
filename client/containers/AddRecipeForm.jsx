import React from 'react';
import axios from 'axios';
import Dropzone from 'react-dropzone';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Loader from 'react-loader';
import { addRecipe } from '../actions/recipe';
import Authenticate from '../auth/auth';

const dropZoneStyles = {
  border: 'none',
  cursor: 'pointer'
};

/**
 * @class AddRecipeForm
 *
 */
export class AddRecipeForm extends React.Component {
  /**
   * @param {obj} props
   * @returns {void} constructor
   */
  constructor(props) {
    super(props);
    this.addRecipe = this.addRecipe.bind(this);
    this.uploadImageToCloudinary = this.uploadImageToCloudinary.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
    this.onChange = this.onChange.bind(this);
    this.state = {
      imageURI: null,
      image: null,
      loaded: true,
      uploadError: null,
      ingredientError: '',
      titleError: '',
      instructionError: '',
      addingRecipe: false
    };
  }

  /**
   * @param {obj} event
   * @returns {void} onChange
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
   *
   * @param {obj} file
   * @return {void} handleDrop
   */
  handleDrop(file) {
    this.setState({ image: file[0] });
  }

  /**
   *
   * @param {obj} event
   * @return {obj} uploadImageToCloudinary
   */
  async uploadImageToCloudinary() {
    try {
      const imageFile = this.state.image;
      const CLOUDINARY_URL = 'https://api.cloudinary.com/v1_1/dsj9ygnq2/upload';
      const CLOUDINARY_UPLOAD_PRESET = 'xmklgrkm';
      const formData = new FormData();
      formData.append('file', imageFile);
      formData.append('upload_preset', CLOUDINARY_UPLOAD_PRESET);
      const response = await axios({
        url: CLOUDINARY_URL,
        method: 'POST',
        headers: { 'X-Requested-With': 'XMLHttpRequest' },
        data: formData
      });
      return Promise.resolve(response.data.secure_url);
    } catch (err) {
      return Promise.reject(err);
    }
  }

  /**
   * @param {obj} event
   * @returns {void} addRecipe
   */
  async addRecipe(event) {
    try {
      event.preventDefault();
      let recipe = {
        title: this.state.title,
        ingredients: this.state.ingredients,
        instructions: this.state.instructions,
      };

      const errors = Authenticate.validateAddRecipe(recipe);

      if (errors.title !== '' || errors.ingredients !== '' ||
     errors.instructions) {
        return this.setState({
          ingredientError: errors.ingredients,
          titleError: errors.title,
          instructionError: errors.instructions
        });
      }
      this.setState({ loaded: false, addingRecipe: true });
      const imageURI = await this.uploadImageToCloudinary();

      recipe = {
        title: this.state.title,
        ingredients: this.state.ingredients,
        instructions: this.state.instructions,
        image: imageURI
      };

      await this.props.addRecipe(recipe);

      this.props.history.push('/view_recipes');
    } catch (err) {
      if (err.response.data.error.message) {
        this.setState({
          loaded: true,
          uploadError: 'image is required',
          addingRecipe: false,
          ingredientError: '',
          titleError: '',
          instructionError: ''
        });
      }
    }
  }

  /**
   * @returns {obj} render
   */
  render() {
    let imageSource = null;
    if (this.state.image) {
      imageSource = this.state.image.preview;
    }
    const { uploadError } = this.state;
    return (
      <div className="main-userboard-body add-recipe-body">
        <div className="container">
          <div className="row">
            <div className="mx-auto col-sm-12 col-md-8 col-lg-8 col-xs mt-5">
              <h2 className="text-center"> Add Recipe</h2>
              <form onSubmit={this.addRecipe}>
                <Dropzone
                  onDrop={this.handleDrop}
                  multiple
                  accept="image/*"
                  style={dropZoneStyles}
                >

                  {this.state.image ?
                    <img className="card-img-top" style={{ height: '450px' }} alt="" src={imageSource} /> :
                    <div className="upload-recipe-img">
                      <div className="row justify-content-center">
                        <div className="col-12">
                          <p className="text-center" id="addImage">
                            <span className="h2"><i className="ion ion-camera" /></span>
                            <br />
                         Click to upload image
                          </p>
                        </div>
                      </div>
                    </div>
                 }
                </Dropzone>
                <div>
                  { uploadError &&
                  <span className="help-block error text-danger">
                    {uploadError}
                  </span>
                    }
                </div>

                <img src={`${this.state.imageURI}`} id="imagePreview" alt="" />
                <div className="form-group">
                  <label htmlFor="food">Recipe Name</label>
                  <input
                    onChange={this.onChange}
                    type="text"
                    name="title"
                    className="form-control form-control-lg"
                    placeholder="Enter Recipe name"
                  />
                  { this.state.titleError &&
                  <span className="help-block error text-danger">
                    {this.state.titleError}
                  </span>
                    }
                </div>

                <div className="form-group">
                  <label htmlFor="ingredient">Ingredients</label>
                  <input
                    onChange={this.onChange}
                    type="text"
                    name="ingredients"
                    className="form-control form-control-lg"
                    placeholder="Enter Ingredients"
                    id="ingredients"
                  />
                  { this.state.ingredientError &&
                    <span className="help-block error text-danger">
                      { this.state.ingredientError}
                    </span>
                    }
                </div>

                <div className="form-group">
                  <label htmlFor="instruction">Instructions</label>
                  <textarea
                    onChange={this.onChange}
                    className="form-control form-control-lg"
                    placeholder="Enter Instructions"
                    name="instructions"
                    id="instructions"
                  />
                  { this.state.instructionError &&
                  <span className="help-block error text-danger">
                    { this.state.instructionError}
                  </span>
                    }
                </div>
                <Loader
                  loaded={this.state.loaded}
                  lines={13}
                  length={30}
                  width={10}
                  radius={30}
                  corners={1}
                  rotate={0}
                  direction={1}
                  color="#000"
                  speed={2}
                  trail={60}
                  shadow={false}
                  hwaccel={false}
                  className="spinner"
                  zIndex={2e9}
                  top="70%"
                  left="50%"
                  scale={1.00}
                  loadedClassName="loadedContent"
                />

                <div className="form-group">
                  <input
                    type="submit"
                    className="btn btn-success"
                    disabled={this.state.addingRecipe}
                    value={this.state.addingRecipe ? 'Adding Recipe...' :
                     'Add Recipe'}
                  />
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    );
  }
}


AddRecipeForm.defaultProps = {
  addRecipe: {}
};

AddRecipeForm.propTypes = {
  addRecipe: PropTypes.func,
  history: PropTypes.shape({
    push: PropTypes.func
  }).isRequired
};


const mapStateToProps = state => ({
  recipes: state.recipes,
  loaded: state.loaded
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ addRecipe }, dispatch);

const ConnectedAddRecipeForm =
  connect(mapStateToProps, mapDispatchToProps)(AddRecipeForm);

export default ConnectedAddRecipeForm;
