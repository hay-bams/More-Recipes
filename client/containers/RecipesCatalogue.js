import React from 'react';
import { Link } from 'react-router-dom';
import _ from 'lodash';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getAllRecipes }  from '../actions/actions.js';

/**
 * @class RecipeCatalogue
 */
class RecipeCatalogue extends React.Component {
  componentDidMount() {
    this.props.getAllRecipes()
  }

  renderCatalogue() {
    const allRecipes = _.mapKeys(this.props.recipes.data, 'id');
    return _.map(allRecipes, recipe => {
      return (
        <div className = "col-12 col-sm-6 col-md-4 col-lg-3 recipes" key={recipe.id}>
          <div className="card recipe-card" style={{border: 'none'}}>
            <img className="card-img-top img-fluid" src="images/recipe5.jpg" alt="Card image cap" style = {{height: 200 + 'px'}} />
            <div className="card-body">
              <h4 className="card-title">{recipe.title}</h4>
              <span className="card-text">
                <span className = "row">
                  <a href="#" className = "text-success">
                  <i className="fa fa-thumbs-up col-4" aria-hidden="true"></i>{recipe.upvotes}</a>
                  <a href="#" className = "text-info"><i className="fa fa-comment col-4" aria-hidden="true"></i>25</a>
                  <a href="#" className = "text-danger"><i className="fa fa-thumbs-down col-4" aria-hidden="true"></i>{recipe.downvotes}</a>
                </span>
              </span>
              <Link to='/details'  className="btn btn-info">view details</Link>
            </div>
          </div>
        </div>
      )
    })
  }

  render() {
      return (
        <div className = "container">
        <div className = "row">
          <div className = "col-sm">
            <h2 className = "text-center catalogue">Recipes Catalogue</h2>
          </div>
        </div>

        <div className = "row">
         {this.renderCatalogue()}
        </div>
      </div>
      )
  }
}

const mapStateToProps = state => {
  return {
    recipes: state.recipes
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ getAllRecipes }, dispatch);
}
export default connect(mapStateToProps, mapDispatchToProps)(RecipeCatalogue);
