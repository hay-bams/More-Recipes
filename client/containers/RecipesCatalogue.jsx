import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { bindActionCreators } from 'redux';
import { getAllRecipes } from '../actions/actions';

import recipeImage from '../images/recipe24.jpg';

/**
 * @class RecipeCatalogue
 */
class RecipeCatalogue extends React.Component {
  /**
   * @returns {void} constructor
   */
  constructor() {
    super();
    this.handlePageClick = this.handlePageClick.bind(this);
    this.renderCatalogue = this.renderCatalogue.bind(this);
  }
  /**
   * @return {void} componentDidMount
   */
  componentDidMount() {
    this.props.getAllRecipes(1);
  }

  /**
   * @param {obj} data
   * @returns {void} handlePageClick
   */
  handlePageClick(data) {
    const selected = data.selected + 1;
    this.props.getAllRecipes(selected);
  }

  /**
   * @returns {obj} renderCatalogue
   */
  renderCatalogue() {
    const { recipes } = this.props;
    return recipes.map(recipe => (
      <div
        className="col-12 col-sm-6 col-md-4 col-lg-3 recipes"
        key={recipe.id}
      >
        <div className="card recipe-card" style={{ border: 'none' }}>
          <h4 className="card-header headerColor text-center">{recipe.title}</h4>
          <img
            className="card-img-top img-fluid img-recipe"
            src={recipe.image}
            alt="Card  cap"
            style={{ height: `${200}px` }}
          />
          <div className="card-body">
          <p>{recipe.instructions.slice(1, 60)}</p>
            <span className="card-text">
              <span className="row ml-1">
                <span className="text-success">
                  <i
                    className="fa fa-thumbs-up col-4"
                    aria-hidden="true"
                  />{recipe.upvotes}
                </span>
                <span className="text-info">
                  <i className="fa fa-comment col-4" aria-hidden="true" />25
                </span>
                <span className="text-danger">
                  <i
                    className="fa fa-thumbs-down col-4"
                    aria-hidden="true"
                  />{recipe.downvotes}
                </span>
              </span>
            </span>
            <Link to={`/details/${recipe.id}`} className="btn btn-outline-info ml-4">view details</Link>
          </div>
        </div>
      </div>
    ));
  }

  /**
   * @returns {obj} render
   */
  render() {
    return (
      <div className="container">
        <div className="row">
          <div className="col-sm">
            <h2 className="text-center catalogue">Recipes Catalogue</h2>
          </div>
        </div>

        <div className="row">
          {this.renderCatalogue()}
        </div>

        <div className="row">
          <div className="col-md-8 ml-auto">
            <ReactPaginate
              previousLabel="Previous"
              nextLabel="Next"
              breakLabel={<a href="">...</a>}
              breakClassName="page-link"
              pageCount={this.props.pages}
              onPageChange={this.handlePageClick}
              containerClassName="pagination pagination-md"
              pageLinkClassName="page-link"
              nextLinkClassName="page-link"
              previousLinkClassName="page-link"
              disabledClassName="disabled"
              pageClassName="page-item"
              previousClassName="page-item"
              nextClassName="page-item"
              activeClassName="active"
              subContainerClassName="pages pagination"
            />
          </div>
        </div>
      </div>
    );
  }
}

RecipeCatalogue.propTypes = {
  getAllRecipes: PropTypes.func.isRequired,
  recipes: PropTypes.arrayOf(PropTypes.shape({
    upvotes: PropTypes.number,
    downvotes: PropTypes.number,
    id: PropTypes.number,
    title: PropTypes.string,
    image: PropTypes.string,
    instructions: PropTypes.string,
    ingredients: PropTypes.string,
    userId: PropTypes.number,
    createdAt: PropTypes.string,
    updatedAt: PropTypes.string
  })).isRequired
};

const mapStateToProps = state => ({
  recipes: state.recipes.rows,
  pages: state.recipes.pages
});

const mapDispatchToProps = dispatch => bindActionCreators(
  { getAllRecipes },
  dispatch
);
export default connect(mapStateToProps, mapDispatchToProps)(RecipeCatalogue);
