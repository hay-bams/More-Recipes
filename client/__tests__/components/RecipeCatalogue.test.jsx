import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import { RecipeCatalogue } from '../../containers/RecipesCatalogue';

describe('Latest Recipe Component', () => {
  const props = {
    getAllRecipes: jest.fn(),
    searchRecipes: jest.fn(),
    recipes: [
      {
        upvotes: 0,
        downvotes: 0,
        id: 1,
        title: 'titlw',
        image: 'image',
        instructions: 'instructions',
        ingredients: 'ingredients',
        userId: 1,
        createdAt: 'created date',
        updatedAt: 'updated date'
      }
    ]
  };

  test('it should render recipeCatalogue correctly', () => {
    const wrapper = shallow(<RecipeCatalogue {...props} />);
    expect(wrapper.find('RecipeCard').exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
 
  test('should set recipeSearch state on input change', () => {
    const value = 'pepper';
    const name = 'recipeSearch';
    const wrapper = shallow(<RecipeCatalogue {...props} />);
    wrapper.find('input').simulate('change', {
      target: { name, value },
      preventDefault: () => {}
    });
    expect(wrapper.state('recipeSearch')).toBe(value);
  });

  test('should call searchRecipes prop on change', () => {
    const value = 'pepper';
    const name = 'recipeSearch';
    const wrapper = shallow(<RecipeCatalogue {...props} />);
    wrapper.instance().setState({
      page: 1,
      sort: 'upvotes',
      order: 'asc'
    });
    wrapper.update();
    wrapper.find('input').simulate('onchange', {
      target: { name, value },
      preventDefault: () => {}
    });
  
    expect(props.searchRecipes).toHaveBeenCalledTimes(1);
  });

  test('it should call onSelect when sort changes', () => {
    const name = 'sort';
    const value = 'upvotes';
    const spy = sinon.spy(RecipeCatalogue.prototype, 'onSelect');
    const wrapper = shallow(<RecipeCatalogue {...props} />);
    wrapper.find('.form-group.col-sm-6').at(0).simulate('change', {
      target: { name, value },
      preventDefault: () => {}
    });
    expect(spy.called).toBeTruthy();
    RecipeCatalogue.prototype.onSelect.restore()
  });

  test('it should call onSelect when order changes', () => {
    const name = 'order';
    const value = 'asc';
    const spy = sinon.spy(RecipeCatalogue.prototype, 'onSelect');
    const wrapper = shallow(<RecipeCatalogue {...props} />);
    wrapper.find('.form-group.col-sm-6').at(1).simulate('change', {
      target: { name, value },
      preventDefault: () => {}
    });
    expect(spy.called).toBeTruthy();
  });
});
 