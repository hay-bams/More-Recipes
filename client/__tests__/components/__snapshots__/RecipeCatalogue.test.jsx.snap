// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Latest Recipe Component it should render recipeCatalogue correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-sm"
    >
      <h2
        className="text-center catalogue"
      >
        Recipe Catalogues
      </h2>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-lg-12"
    >
      <form>
        <div
          className="form-group"
        >
          <input
            className="form-control form-control-lg"
            name="recipeSearch"
            onChange={[Function]}
            placeholder="Search by ingredients or name"
            type="search"
          />
          <hr />
          <div
            className="row"
          >
            <div
              className="form-group col-sm-6"
              onChange={[Function]}
            >
              <select
                className="form-control"
                name="sort"
              >
                <option
                  value=""
                >
                  Sort By
                </option>
                <option
                  value="upvotes"
                >
                  Upvotes
                </option>
                <option
                  value="downvotes"
                >
                  Downvotes
                </option>
              </select>
            </div>
            <div
              className="form-group col-sm-6"
              onChange={[Function]}
            >
              <select
                className="form-control"
                name="order"
              >
                <option
                  value=""
                >
                  Order By
                </option>
                <option
                  value="asc"
                >
                  Ascending
                </option>
                <option
                  value="desc"
                >
                  Descending
                </option>
              </select>
            </div>
          </div>
        </div>
      </form>
    </div>
    <RecipeCard
      createdAt="created date"
      downvotes={0}
      id={1}
      image="image"
      ingredients="ingredients"
      instructions="instructions"
      key="1"
      title="titlw"
      updatedAt="updated date"
      upvotes={0}
      userId={1}
    />
  </div>
  <Component
    className="spinner"
    color="#000"
    component="div"
    corners={1}
    direction={1}
    hwaccel={false}
    left="50%"
    length={30}
    lines={13}
    loaded={false}
    loadedClassName="loadedContent"
    parentClassName="loader"
    radius={30}
    rotate={0}
    scale={1}
    shadow={false}
    speed={2}
    top="70%"
    trail={60}
    width={10}
    zIndex={2000000000}
  />
  <div
    className="row"
  >
    <div
      className="col-md-8 ml-auto"
    >
      <PaginationBoxView
        activeClassName="active"
        breakClassName="page-link"
        breakLabel={
          <a
            href=""
          >
            ...
          </a>
        }
        containerClassName="pagination pagination-md"
        disableInitialCallback={false}
        disabledClassName="disabled"
        marginPagesDisplayed={3}
        nextClassName="page-item"
        nextLabel="Next"
        nextLinkClassName="page-link"
        onPageChange={[Function]}
        pageClassName="page-item"
        pageCount={10}
        pageLinkClassName="page-link"
        pageRangeDisplayed={2}
        previousClassName="page-item"
        previousLabel="Previous"
        previousLinkClassName="page-link"
        subContainerClassName="pages pagination"
      />
    </div>
  </div>
</div>
`;
